raiden/messages/transfers.py

```diff
@@ -406,19 +412,95 @@ def from_event(cls, event: Any) -> Any:  # noqa: F811
         )
 
     def _packed_data(self) -> bytes:
+        return LockedTransferBase._pack_locked_transfer_data(
+            self.cmdid,  # pylint: disable=no-member
+            self.message_identifier,
+            self.payment_identifier,
+            self.token,
+            self.recipient,
+            self.target,
+            self.initiator,
+            self.lock,
+        )
+
+    @classmethod
+    def _pack_locked_transfer_data(
+        cls,
+        cmdid: CmdId,
+        message_identifier: MessageID,
+        payment_identifier: PaymentID,
+        token: TokenAddress,
+        recipient: Address,
+        target: TargetAddress,
+        initiator: InitiatorAddress,
+        lock: Lock,
+    ) -> bytes:
         return (
-            bytes([self.cmdid.value])  # pylint: disable=no-member
-            + self.message_identifier.to_bytes(8, byteorder="big")
-            + self.payment_identifier.to_bytes(8, byteorder="big")
-            + self.lock.expiration.to_bytes(32, byteorder="big")
-            + self.token
-            + self.recipient
-            + self.target
-            + self.initiator
-            + self.lock.secrethash
-            + self.lock.amount.to_bytes(32, byteorder="big")
+            bytes([cmdid.value])
+            + message_identifier.to_bytes(8, byteorder="big")
+            + payment_identifier.to_bytes(8, byteorder="big")
+            + lock.expiration.to_bytes(32, byteorder="big")
+            + token
+            + recipient
+            + target
+            + initiator
+            + lock.secrethash
+            + lock.amount.to_bytes(32, byteorder="big")
+        )
+
+    @post_load(pass_original=True)
+    def _check_metadata_hash(  # pylint: disable=no-self-use,unused-argument
+        self, data: Dict[str, Any], original_data: Dict[str, Any], many: bool, **kwargs: Any
+    ) -> Dict[str, Any]:
+
+        metadata = data["metadata"]
```

Comment:
Regarding the chosen-hash attack mentioned above, to mitigate you could check for existence of `metadata['_hash']` field here and drop the message somehow?. It is a little hacky but would maybe do the trick...

Reply:
The only thing that's still a little annoying with this approach is that `_legacy_hash` is now officially a reserved attribute that can't be used in the arbitrarily definable metadata, just because of internal implementation details.
This is not a big issue though.

Also, I would feel more comfortable if you still explicitly drop messages with `_legacy_hash` present, since there is still some entropy left that the sender can chose for validation.